edge-util/src/Microsoft.Azure.Devices.Edge.Storage/DbStoreProviderWithBackupRestore.cs

```diff
@@ -0,0 +1,353 @@
+            IDbStoreBackupRestore dbStoreBackupRestore,
+            SerializationFormat backupFormat)
+        {
+            DbStoreProviderWithBackupRestore provider = new DbStoreProviderWithBackupRestore(dbStoreProvider, backupPath, dbStoreBackupRestore, backupFormat);
+            await provider.RestoreAsync();
+            return provider;
+        }
+
+        private async Task RestoreAsync()
+        {
+            string backupMetadataFilePath = Path.Combine(this.backupPath, BackupMetadataFileName);
+
+            if (!File.Exists(backupMetadataFilePath))
+            {
+                this.events.NoBackupsForRestore();
+                return;
+            }
+
+            try
+            {
+                string fileText = File.ReadAllText(backupMetadataFilePath);
+                BackupMetadataList backupMetadataList = JsonConvert.DeserializeObject<BackupMetadataList>(fileText);
+                BackupMetadata backupMetadata = backupMetadataList.Backups[0];
+                this.events.BackupInformation(backupMetadata.Id, backupMetadata.SerializationFormat, backupMetadata.TimestampUtc, backupMetadata.Stores);
+
+                string latestBackupDirPath = Path.Combine(this.backupPath, backupMetadata.Id.ToString());
+                if (Directory.Exists(latestBackupDirPath))
+                {
+                    this.events.RestoringFromBackup(backupMetadata.Id);
+                    foreach (string store in backupMetadata.Stores)
+                    {
+                        IDbStore dbStore;
+                        if (!store.Equals(DefaultStoreBackupName, StringComparison.OrdinalIgnoreCase))
+                        {
+                            dbStore = this.GetDbStore(store);
+                        }
+                        else
+                        {
+                            dbStore = this.GetDbStore();
+                        }
+
+                        await this.dbStoreBackupRestore.RestoreAsync(store, dbStore, latestBackupDirPath);
+                    }
+
+                    this.events.RestoreComplete();
+                }
+                else
+                {
+                    this.events.NoBackupsForRestore();
+                }
+
+                this.CleanupAllBackups(this.backupPath);
+            }
+            catch (Exception exception) when (
+            exception is IOException
+            || exception is JsonException)
+            {
+                this.events.RestoreFailure($"The restore operation failed with error ${exception}.");
+
+                // Clean up any restored state in the dictionary if the backup fails midway.
+                foreach (string store in this.dbStores.ToList())
+                {
+                    this.RemoveDbStore(store);
+                }
+
+                this.RemoveDbStore();
+
+                // Delete all backups as the last backup itself is corrupt.
+                this.CleanupAllBackups(this.backupPath);
```

Comment:
Maybe it is valid to clean up after a non IO/JSonException, too, and the CleanupAllBackups() could go into finaly?

Reply:
Yeah, sounds good. I'll change it

Reply:
Yeah, sounds good. I'll change it